src/common/test-json.c

```diff
@@ -676,6 +676,155 @@ test_write_infinite_nan (void)
   g_free (string);
 }
 
+static JsonNode *
+flip_integer (JsonNode *node,
+              gpointer  user_data)
+{
+  int *counter = user_data;
+
+  if (json_node_get_value_type (node) != G_TYPE_INT64)
+    return NULL;
+
+  int value = json_node_get_int (node);
+
+  (*counter)++;
+
+  return json_node_init_int (json_node_alloc(), -value);
+}
+
+/* replace all three-character strings with the word 'three' */
+static JsonNode *
+three_subst (JsonNode *node,
+             gpointer  user_data)
+{
+  int *counter = user_data;
+
+  const gchar *value = json_node_get_string (node);
+
+  if (value == NULL || strlen (value) != 3)
+    return NULL;
+
+  (*counter)++;
+
+  return json_node_init_string (json_node_alloc(), "three");
+}
+
+static JsonNode *
+null_subst (JsonNode *node,
+            gpointer  user_data)
+{
+  int *counter = user_data;
+
+  (*counter)++;
+
+  return NULL;
+}
+
+static void
+test_walk (void)
```

Comment:
This looks pretty good to me, and with valgrind this should ensure that the refcounting works. I think the only missing case is to cover JSON `null`s, as above.

Reply:
We do actually test that null is handled properly since the existing testcases for `cockpit_json_patch()` have a `null`, and we reuse that data...

Reply:
Was missing that, thanks!